### ISSUE TITLE
Text field data binded to parent view instead of the current.

### ISSUE THREAD

**Issue by santiagopuentep** [ManuelDeLeon/viewmodel#102]
I have an `item` template that has a text field with a data bind to a getText function and is rendered by an `each` statement and that template has many nested children.

```
{{#each items}}
    {{> item}}
{{/each}}
```
Each item is like this:

```
<template name="item">
	<div class="textArea" contenteditable="true" data-bind="text: getText">
	{{#if hasChildren}}
		{{#unless getCollapsed}}
			<div class="itemContainer">
				{{#each getChildren}}
					{{> item}}
				{{/each}}
			</div>
		{{/unless}}
	{{/if}}
</template>
```

The `getText()` helper gets the text data from a Collection using `data._id` and a `find`.

The problem is that the `data` context inside the `getText` helper for the children is the root item's data for some reason. The onRendered helpers of the children get called after the function has been called and the data context there is correct.

Is maybe the data-bind not finding the `getText` function and searching up until it finds it in the parent?

Sometimes I even get an error in the root item because its rendered before the data context is the correct one. It seems to have something to do with the root `#each` statement.

**Comment by ManuelDeLeon** [comment]
I need a repro for this...

**Comment by santiagopuentep** [comment]
I'll build one now.

**Comment by santiagopuentep** [comment]
Here's the reproduction:

The child templates get bound to the parent instead of themselves. In the faq there's something writen about the problems with the `if` and `each` statements but I'm using templates and I don't know how to go around this.

https://github.com/santiagopuentep/viewmodel-bind-problem

I'm using an `each` in the root because I had problems with a `with` that I talked about in my last ticket:

https://github.com/ManuelDeLeon/viewmodel/issues/89#issuecomment-130455363

**Comment by ManuelDeLeon** [comment]
I don't understand the problem. 

Given the data:
![image](https://cloud.githubusercontent.com/assets/4257750/9641879/4788601c-5176-11e5-8230-2e0f9eed3015.png)

This is the output:
![image](https://cloud.githubusercontent.com/assets/4257750/9641891/59b39a9a-5176-11e5-88c7-ab2e30ea378a.png)

Which as far as I can tell it the right result. What am I missing?

**Comment by santiagopuentep** [comment]
On my end I get "Root" as text for all the items, but not always, sometimes the result is correct but most of the time it's not.

Please try refreshing the browser a few times.

**Comment by santiagopuentep** [comment]
![2015-09-02_163918](https://cloud.githubusercontent.com/assets/137173/9642208/328cd826-5191-11e5-8c60-0e29588f2439.jpg)

I tried it again and I almost always get the correct result but sometimes not.

**Comment by ManuelDeLeon** [comment]
This is most likely because Blaze can call any onRendered function at any order. I have monkey patched Meteor for it but it could be that it's not foolproof. Try the following:

```html
<div class="desk">
      {{#if root}}
        {{> item root}}
      {{/if}}
</div>
```

```js
root: function() {
  return ItemsList.findOne({root: { $exists: true } });
}
```

If that doesn't work then you need to wait for the subscription to be ready before displaying the items.

**Comment by santiagopuentep** [comment]
That works, but it's the same as using `with`, and I'm back to the problem I had in the other ticket https://github.com/ManuelDeLeon/viewmodel/issues/89

I get no reactivity when the root value changes because the viewmodel is not recreated. I'm stuck...

Is there a way to fix it without resorting to tricks like this one or is Blaze is not ready for production yet? Too many bugs like this one and I had to patch many times already... It's a shame because I love viewmodel. I really don't want to use React or Angular!

**Comment by ManuelDeLeon** [comment]
They don't see it as a bug in Blaze and technically it's not. It's just inconvenient the way it calls the onRendered event.

Did you try waiting for the subscription to be ready?

**Comment by ManuelDeLeon** [comment]
> That works, but it's the same as using with, and I'm back to the problem I had in the other ticket #89

Now that I think of it, one of the last few updates should have resolved that issue. Try it with the `{{#if root }}` and `{{> item root}}` as in my example.

**Comment by santiagopuentep** [comment]
The `#if` works with this data-binding bug but it doesn't work with the problem I had in the other ticket. The template doesn't get recreated on data context change in the `if` or `with`. If I use the `each` statement on a cursor that only returns the root I get the functionality I want but I get this data-binding bug sometimes.

Using an `#each` and waiting for the subscriptions to be ready solved the problem. I don't really understand why the items have their text content bound to the root data. If the child items are being created doesn't it mean that the subscription is ready? The root helper is getting the root data so the subscription should be there already right?

**Comment by ManuelDeLeon** [comment]
Most of the time Blaze calls the onRendered of the children before calling the parent's. The bind occurs in the onRendered. I monkey patch Blaze so that happens *almost* always happens. You've managed to create a (very convoluted) scenario where the parent's onRendered is called before the children's, even though the children are already rendered on screen. 

Again, technically it's not a bug in Blaze. Just a really inconvenient behavior.

> The template doesn't get recreated on data context change in the if or with. 

But the fields that depend on the context do change. e.g. If root is `{ _id: 1, letter: "A" }` and you have

```js
      {{#if root}}
        {{> item root}}
      {{/if}}
```

The the `item` view model will have the properties `_id` (1) and `letter` ("A").

If root changes to `{ _id: 2, letter: "B" }` then the `item` view model will have the properties `_id` (2) and `letter` ("B").

If that's not enough for you then I suggest you rethink what you're doing because it's probably needlessly convoluted.

**Comment by santiagopuentep** [comment]
The root simply needs to represent an item with an id. The problem is that I need to be able to search through the viewmodels using the viewmodel name. Since the root doesn't get recreated when the with context changes the old viewmodel stays there.

My app is not convoluted at all, it's simply items that can have child items nested in them and the way I set them up is simply the intuitive one, based on a collection. I think the problems like not being able to bind in `#if` and the invonvenient order of the onRendered callbacks are really troublesome. If Blaze/viewmodel wants to get big those things I think should be the first ones to fix. I took my time to try to solve this but many people won't.

I think Blaze is a great idea an viewmodel patches almost all of it's problems and adds tons of great features, but I'm really looking forward to the next versions and I hope they solve all these issues. I don't understand for example how the child/parent relationships between templates aren't implemented in Blaze from the start... viewmodel saves the day.

Thanks a lot for the help! Now I can continue with my project.
